File: Arrays-And-Strings/Challenge-5.js
// O(Max(N, M)) time complexity and O(N+M) space complexity

const oneAway = (string1, string2) => {
  if (Math.abs(string1.length - string2.length) > 1) {
    return false;
  }
  const hashMap = {};
  let extraCharacterCount = 0;

  for (let i = 0; i < string1.length; i++) {
    const character = string1[i];

    if (hashMap[character]) {
      hashMap[character] += 1;
    } else {
      hashMap[character] = 1;
    }
  }

  for (let i = 0; i < string2.length; i++) {
    const character = string2[i];

    if (hashMap[character]) {
      hashMap[character] -= 1;
    } else {
      extraCharacterCount += 1;
    }
  }
  let count = 0;
  const characterValues = Object.values(hashMap);

  for (let i = 0; i < characterValues.length; i++) {
    const value = characterValues[i];

    count += value;
  }

  return count + extraCharacterCount <= 2;
};

console.log(oneAway('', '')); // true
console.log(oneAway('abc', 'ab')); // true
console.log(oneAway('aaa', 'bbb')); // false
console.log(oneAway('abc', 'abcdfe')); // false
console.log(oneAway('sale, ', 'sle')); // true
console.log(oneAway('melon', 'felon')); // true
console.log(oneAway('as', 'asdf')); // false

File: Arrays-And-Strings/Challenge-5.py
def one_away(string1,string2):
  if abs(len(string1) - len(string2) > 1):
    return False

  hash_map = {}
  extra_character_count = 0

  for char in string1:
    if hash_map.get(char):
      hash_map[char] += 1
    else:
      hash_map[char] = 1

  for char in string2:
    if hash_map.get(char):
      hash_map[char] -= 1
    else:
      extra_character_count += 1

  count = 0
  character_values = list(hash_map.values())

  for value in character_values:
    count += value

  return count + extra_character_count <= 2

print(one_away('abc', 'ab')) # True
print(one_away('aaa', 'bbb')) # False
